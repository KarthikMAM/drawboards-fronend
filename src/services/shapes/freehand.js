import update from 'immutability-helper'

import { READY, CREATE, RECORDING, FINISH } from '../../constants/states'
import { INIT, MOUSEMOVE, MOUSEDOWN, MOUSEUP, MOUSELEAVE } from '../../constants/events'

export const draw = (data, canvasContext, width, height) => {
  canvasContext.strokeStyle = data.stroke || 'black'
  data.points.forEach((point, index) => {
    if (index === 0) {
      canvasContext.moveTo(point.x * width, point.y * height)
    }

    canvasContext.lineTo(point.x * width, point.y * height)
  })
  canvasContext.stroke()
}

export const transitions = {
  [INIT]: {
    type: INIT,
    transitions: {
      [CREATE]: READY,
    },
    reduce: (data, action) => data,
  },
  [MOUSEMOVE]: {
    type: MOUSEMOVE,
    transitions: {
      [RECORDING]: RECORDING,
    },
    reduce: (data, action) => update(data, {
      points: {
        $push: [action.payload],
      },
    }),
  },
  [MOUSEDOWN]: {
    type: MOUSEDOWN,
    transitions: {
      [READY]: RECORDING,
    },
    reduce: (data, action) => update(data, {
      points: {
        $push: [action.payload],
      },
    }),
  },
  [MOUSEUP]: {
    type: MOUSEUP,
    transitions: {
      [RECORDING]: FINISH,
    },
    reduce: (data, action) => update(data, {
      points: {
        $push: [action.payload],
      },
    }),
  },
  [MOUSELEAVE]: {
    type: MOUSELEAVE,
    transitions: {
      [RECORDING]: FINISH,
    },
    reduce: (data, action) => update(data, {
      points: {
        $push: [action.payload],
      },
    }),
  },

}
