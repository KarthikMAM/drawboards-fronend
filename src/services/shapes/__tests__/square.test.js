import { draw, transitions } from '../freehand'

describe('freehand', () => {
  describe('draw', () => {
    it('should draw all points properly with scaling', () => {
      const contextMock = {
        moveTo: jest.fn(),
        lineTo: jest.fn(),
        stroke: jest.fn(),
      }

      draw({
        stroke: 'green',
        points: [
          { x: 0.2, y: 0.3 },
          { x: 0.32, y: 0.43 },
          { x: 0.34, y: 0.44 },
        ],
      }, contextMock, 100, 200)

      expect(contextMock.strokeStyle).toBe('green')
      expect(contextMock.moveTo).toBeCalledTimes(1)
      expect(contextMock.lineTo).toBeCalledTimes(3)
      expect(contextMock.stroke).toBeCalledTimes(1)
      expect(contextMock).toMatchSnapshot()
    })
  })

  describe('transitions', () => {
    describe('reduce', () => {
      it('should return data as it is during the init phase', () => {
        const data = { hello: 'world' }

        expect(transitions['init'].reduce(data, { action: { type: 'init' } })).toBe(data)
      })

      it('should append the payload to the end during mousemove, mouseup and mouseleave events', () => {
        const data = { points: [{ x: 0.5, y: 0.6 }] }

        expect(transitions['mousemove'].reduce(data, { action: { type: 'init', payload: { x: 1, y: 0.4 } } })).toMatchSnapshot()
        expect(transitions['mousedown'].reduce(data, { action: { type: 'init', payload: { x: 1, y: 0.4 } } })).toMatchSnapshot()
        expect(transitions['mouseup'].reduce(data, { action: { type: 'init', payload: { x: 1, y: 0.4 } } })).toMatchSnapshot()
        expect(transitions['mouseleave'].reduce(data, { action: { type: 'init', payload: { x: 1, y: 0.4 } } })).toMatchSnapshot()
      })
    })

    describe('lifecycle transitions', () => {
      expect(Object.keys(transitions).reduce((acc, value) => {
        acc[value] = transitions[value].transitions
        return acc
      }, {})).toMatchSnapshot()
    })
  })
})
