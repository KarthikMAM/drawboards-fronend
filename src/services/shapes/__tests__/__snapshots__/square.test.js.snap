// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` 1`] = `
Object {
  "init": Object {
    "@states/CREATE": "@states/READY",
  },
  "mousedown": Object {
    "@states/READY": "@states/RECORDING",
  },
  "mouseleave": Object {
    "@states/RECORDING": "@states/FINISH",
  },
  "mousemove": Object {
    "@states/RECORDING": "@states/RECORDING",
  },
  "mouseup": Object {
    "@states/RECORDING": "@states/FINISH",
  },
}
`;

exports[`freehand draw should draw all points properly with scaling 1`] = `
Object {
  "lineTo": [MockFunction] {
    "calls": Array [
      Array [
        20,
        60,
      ],
      Array [
        32,
        86,
      ],
      Array [
        34,
        88,
      ],
    ],
    "results": Array [
      Object {
        "isThrow": false,
        "value": undefined,
      },
      Object {
        "isThrow": false,
        "value": undefined,
      },
      Object {
        "isThrow": false,
        "value": undefined,
      },
    ],
  },
  "moveTo": [MockFunction] {
    "calls": Array [
      Array [
        20,
        60,
      ],
    ],
    "results": Array [
      Object {
        "isThrow": false,
        "value": undefined,
      },
    ],
  },
  "stroke": [MockFunction] {
    "calls": Array [
      Array [],
    ],
    "results": Array [
      Object {
        "isThrow": false,
        "value": undefined,
      },
    ],
  },
  "strokeStyle": "green",
}
`;

exports[`freehand transitions reduce should append the payload to the end during mousemove, mouseup and mouseleave events 1`] = `
Object {
  "points": Array [
    Object {
      "x": 0.5,
      "y": 0.6,
    },
    undefined,
  ],
}
`;

exports[`freehand transitions reduce should append the payload to the end during mousemove, mouseup and mouseleave events 2`] = `
Object {
  "points": Array [
    Object {
      "x": 0.5,
      "y": 0.6,
    },
    undefined,
  ],
}
`;

exports[`freehand transitions reduce should append the payload to the end during mousemove, mouseup and mouseleave events 3`] = `
Object {
  "points": Array [
    Object {
      "x": 0.5,
      "y": 0.6,
    },
    undefined,
  ],
}
`;

exports[`freehand transitions reduce should append the payload to the end during mousemove, mouseup and mouseleave events 4`] = `
Object {
  "points": Array [
    Object {
      "x": 0.5,
      "y": 0.6,
    },
    undefined,
  ],
}
`;
